### ISSUE TITLE
OverflowException in MRefBuilder

### ISSUE THREAD

**Issue by ritchiecarroll** [EWSoftware/SHFB#474]
Any ideas? Having trouble with an enumeration value somewhere?

FYI - this is part of a SandCastle help project in the open source Grid Solutions Framework solution:
https://github.com/GridProtectionAlliance/gsf

MRefBuilder : error : An unexpected error occurred trying to execute the MRefBuilder MSBuild task: System.OverflowException: Value was either too large or too small for an Int64. [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at System.Convert.ToInt64(UInt64 value) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at System.UInt64.System.IConvertible.ToInt64(IFormatProvider provider) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at System.Convert.ToInt64(Object value, IFormatProvider provider) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.GetAppliedFields(EnumNode enumeration, Int64 value) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteLiteral(Literal literal, Boolean showType) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteMember(Member member, TypeNode type, Boolean includeSourceCodeContext) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitMember(Member member) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitMembers(MemberList members) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitType(TypeNode type) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitType(TypeNode type) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitTypes(TypeNodeList types) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitNamespace(Namespace space) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitNamespace(Namespace space) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitNamespaces(NamespaceList spaces) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitNamespaces(NamespaceList spaces) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitApis() [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.MRefBuilderCore.Main(String[] args) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.MSBuild.MRefBuilder.Execute() [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
Done Building Project &quot;D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj&quot; (default targets) -- FAILED.

**Comment by ritchiecarroll** [comment]
Found the culprit - see Bit63 below...

```
    /// <summary>
    /// Represents bits in a signed or unsigned integer value.
    /// </summary>
    [SuppressMessage("Microsoft.Design", "CA1008:EnumsShouldHaveZeroValue"), Flags]
    public enum Bits : ulong
    {
        /// <summary>No bits set (0x0000000000000000)</summary>
        Nil = 0x00000000,

        // Byte 0, Bits 0-7

        /// <summary>Bit 00 (0x0000000000000001)</summary>
        Bit00 = 0x00000001,     // 00000001 = 1

        /// <summary>Bit 01 (0x0000000000000002)</summary>
        Bit01 = Bit00 << 1,     // 00000010 = 2

        /// <summary>Bit 02 (0x0000000000000004)</summary>
        Bit02 = Bit01 << 1,     // 00000100 = 4

        /// <summary>Bit 03 (0x0000000000000008)</summary>
        Bit03 = Bit02 << 1,     // 00001000 = 8

        /// <summary>Bit 04 (0x0000000000000010)</summary>
        Bit04 = Bit03 << 1,     // 00010000 = 16

        /// <summary>Bit 05 (0x0000000000000020)</summary>
        Bit05 = Bit04 << 1,     // 00100000 = 32

        /// <summary>Bit 06 (0x0000000000000040)</summary>
        Bit06 = Bit05 << 1,     // 01000000 = 64

        /// <summary>Bit 07 (0x0000000000000080)</summary>
        Bit07 = Bit06 << 1,     // 10000000 = 128

        // Byte 1, Bits 8-15

        /// <summary>Bit 08 (0x0000000000000100)</summary>
        Bit08 = Bit07 << 1,

        /// <summary>Bit 09 (0x0000000000000200)</summary>
        Bit09 = Bit08 << 1,

        /// <summary>Bit 10 (0x0000000000000400)</summary>
        Bit10 = Bit09 << 1,

        /// <summary>Bit 11 (0x0000000000000800)</summary>
        Bit11 = Bit10 << 1,

        /// <summary>Bit 12 (0x0000000000001000)</summary>
        Bit12 = Bit11 << 1,

        /// <summary>Bit 13 (0x0000000000002000)</summary>
        Bit13 = Bit12 << 1,

        /// <summary>Bit 14 (0x0000000000004000)</summary>
        Bit14 = Bit13 << 1,

        /// <summary>Bit 15 (0x0000000000008000)</summary>
        Bit15 = Bit14 << 1,

        // Byte 2, Bits 16-23

        /// <summary>Bit 16 (0x0000000000010000)</summary>
        Bit16 = Bit15 << 1,

        /// <summary>Bit 17 (0x0000000000020000)</summary>
        Bit17 = Bit16 << 1,

        /// <summary>Bit 18 (0x0000000000040000)</summary>
        Bit18 = Bit17 << 1,

        /// <summary>Bit 19 (0x0000000000080000)</summary>
        Bit19 = Bit18 << 1,

        /// <summary>Bit 20 (0x0000000000100000)</summary>
        Bit20 = Bit19 << 1,

        /// <summary>Bit 21 (0x0000000000200000)</summary>
        Bit21 = Bit20 << 1,

        /// <summary>Bit 22 (0x0000000000400000)</summary>
        Bit22 = Bit21 << 1,

        /// <summary>Bit 23 (0x0000000000800000)</summary>
        Bit23 = Bit22 << 1,

        // Byte 3, Bits 24-31

        /// <summary>Bit 24 (0x0000000001000000)</summary>
        Bit24 = Bit23 << 1,

        /// <summary>Bit 25 (0x0000000002000000)</summary>
        Bit25 = Bit24 << 1,

        /// <summary>Bit 26 (0x0000000004000000)</summary>
        Bit26 = Bit25 << 1,

        /// <summary>Bit 27 (0x0000000008000000)</summary>
        Bit27 = Bit26 << 1,

        /// <summary>Bit 28 (0x0000000010000000)</summary>
        Bit28 = Bit27 << 1,

        /// <summary>Bit 29 (0x0000000020000000)</summary>
        Bit29 = Bit28 << 1,

        /// <summary>Bit 30 (0x0000000040000000)</summary>
        Bit30 = Bit29 << 1,

        /// <summary>Bit 31 (0x0000000080000000)</summary>
        Bit31 = Bit30 << 1,

        // Byte 4, Bits 32-39

        /// <summary>Bit 32 (0x0000000100000000)</summary>
        Bit32 = Bit31 << 1,

        /// <summary>Bit 33 (0x0000000200000000)</summary>
        Bit33 = Bit32 << 1,

        /// <summary>Bit 34 (0x0000000400000000)</summary>
        Bit34 = Bit33 << 1,

        /// <summary>Bit 35 (0x0000000800000000)</summary>
        Bit35 = Bit34 << 1,

        /// <summary>Bit 36 (0x0000001000000000)</summary>
        Bit36 = Bit35 << 1,

        /// <summary>Bit 37 (0x0000002000000000)</summary>
        Bit37 = Bit36 << 1,

        /// <summary>Bit 38 (0x0000004000000000)</summary>
        Bit38 = Bit37 << 1,

        /// <summary>Bit 39 (0x0000008000000000)</summary>
        Bit39 = Bit38 << 1,

        // Byte 5, Bits 40-47

        /// <summary>Bit 40 (0x0000010000000000)</summary>
        Bit40 = Bit39 << 1,

        /// <summary>Bit 41 (0x0000020000000000)</summary>
        Bit41 = Bit40 << 1,

        /// <summary>Bit 42 (0x0000040000000000)</summary>
        Bit42 = Bit41 << 1,

        /// <summary>Bit 43 (0x0000080000000000)</summary>
        Bit43 = Bit42 << 1,

        /// <summary>Bit 44 (0x0000100000000000)</summary>
        Bit44 = Bit43 << 1,

        /// <summary>Bit 45 (0x0000200000000000)</summary>
        Bit45 = Bit44 << 1,

        /// <summary>Bit 46 (0x0000400000000000)</summary>
        Bit46 = Bit45 << 1,

        /// <summary>Bit 47 (0x0000800000000000)</summary>
        Bit47 = Bit46 << 1,

        // Byte 6, Bits 48-55

        /// <summary>Bit 48 (0x0001000000000000)</summary>
        Bit48 = Bit47 << 1,

        /// <summary>Bit 49 (0x0002000000000000)</summary>
        Bit49 = Bit48 << 1,

        /// <summary>Bit 50 (0x0004000000000000)</summary>
        Bit50 = Bit49 << 1,

        /// <summary>Bit 51 (0x0008000000000000)</summary>
        Bit51 = Bit50 << 1,

        /// <summary>Bit 52 (0x0010000000000000)</summary>
        Bit52 = Bit51 << 1,

        /// <summary>Bit 53 (0x0020000000000000)</summary>
        Bit53 = Bit52 << 1,

        /// <summary>Bit 54 (0x0040000000000000)</summary>
        Bit54 = Bit53 << 1,

        /// <summary>Bit 55 (0x0080000000000000)</summary>
        Bit55 = Bit54 << 1,

        // Byte 7, Bits 56-63

        /// <summary>Bit 56 (0x0100000000000000)</summary>
        Bit56 = Bit55 << 1,

        /// <summary>Bit 57 (0x0200000000000000)</summary>
        Bit57 = Bit56 << 1,

        /// <summary>Bit 58 (0x0400000000000000)</summary>
        Bit58 = Bit57 << 1,

        /// <summary>Bit 59 (0x0800000000000000)</summary>
        Bit59 = Bit58 << 1,

        /// <summary>Bit 60 (0x1000000000000000)</summary>
        Bit60 = Bit59 << 1,

        /// <summary>Bit 61 (0x2000000000000000)</summary>
        Bit61 = Bit60 << 1,

        /// <summary>Bit 62 (0x4000000000000000)</summary>
        Bit62 = Bit61 << 1,

        /// <summary>Bit 63 (0x8000000000000000)</summary>
        Bit63 = Bit62 << 1
    }
```

**Comment by ritchiecarroll** [comment]
Error is here:
```
System.Convert.ToInt64(GSF.Bits.Bit63)
'System.Convert.ToInt64(GSF.Bits.Bit63)' threw an exception of type 'System.OverflowException'
    Data: {System.Collections.ListDictionaryInternal}
    HResult: -2146233066
    HelpLink: null
    InnerException: null
    Message: "Value was either too large or too small for an Int64."
    Source: "mscorlib"
    StackTrace: "   at System.Convert.ToInt64(UInt64 value)"
    TargetSite: {Int64 ToInt64(UInt64)}
```

**Comment by EWSoftware** [comment]
See the fix in the [latest commit](https://github.com/EWSoftware/SHFB/commit/a3fd713bbdd734bdcb48260bb16cef283d76ae68#diff-49913d76dfd4ca54ce6ec41788dc7daa).  It fixes this issue as well as one that was causing duplicate values to be returned when they matched those in flag values made up of combined values.

**Comment by ritchiecarroll** [comment]
Not quite - my original pull request handled cases where enum values can be negative:
```
MRefBuilder : error :    at System.Convert.ToUInt64(Int32 value) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at System.Int32.System.IConvertible.ToUInt64(IFormatProvider provider) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at System.Convert.ToUInt64(Object value, IFormatProvider provider) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteLiteral(Literal literal, Boolean showType) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteExpression(Expression expression) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteAttributes(AttributeList attributes, SecurityAttributeList securityAttributes) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteMember(Member member, TypeNode type, Boolean includeSourceCodeContext) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteTypeElements(TypeNode type) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.WriteType(TypeNode type) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitType(TypeNode type) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitTypes(TypeNodeList types) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitNamespace(Namespace space) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitNamespace(Namespace space) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitNamespaces(NamespaceList spaces) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.ManagedReflectionWriter.VisitNamespaces(NamespaceList spaces) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.Reflection.ApiVisitor.VisitApis() [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.MRefBuilderCore.Main(String[] args) [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error :    at Microsoft.Ddue.Tools.MSBuild.MRefBuilder.Execute() [D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj]
Done Building Project &quot;D:\Projects\gsf\Build\Help\Working\GenerateRefInfo.proj&quot; (default targets) -- FAILED.
```
Such as:
```
    /// <summary>
    /// Defines the relative intensity (i.e., volume) of a musical line.
    /// </summary>
    public enum Dynamic
    {
        /// <summary>No dynamic is defined.</summary>
        Undefined = -1,
        /// <summary>pp - very soft.</summary>
        Pianissimo = 5,
        /// <summary>p - soft.</summary>
        Piano = 10,
        /// <summary>mp - half soft as <see cref="Piano"/>.</summary>
        MezzoPiano = 20,
        /// <summary>mf - half loud as <see cref="Forte"/>.</summary>
        /// <remarks>This is the default dynamic level.</remarks>
        MezzoForte = 40,
        /// <summary>f - loud.</summary>
        Forte = 80,
        /// <summary>ff - very loud.</summary>
        Fortissimo = 90
    }
```

**Comment by ritchiecarroll** [comment]
New PR fixes - see #476

**Comment by ritchiecarroll** [comment]
FYI - I cannot re-open issue that owner has closed. Thanks!
